cmd/config/crypto.go

```diff
@@ -0,0 +1,323 @@
+	return KeyDerivationParams{
+		Algorithm:  "Argon2id",
+		Salt:       salt,
+		TimeCost:   1,
+		MemoryCost: 64 * 1024,
+		Threads:    4,
+	}
+}
+
+// MasterKey is a 256 bit cryptographic key that can be used to
+// encrypt data streams.
+type MasterKey []byte
+
+// DeriveMasterKey derives a new master key from the given password
+// and key deriviation parameters.
+//
+// In the context of S3, the password should be the S3 secret key.
+//
+// It returns an error if the key derivation parameters are invalid
+// or would cause a very expensive key derivation.
+func DeriveMasterKey(password string, params KeyDerivationParams) (MasterKey, error) {
+	// Define some reasonable max. values. We should not
+	// accept arbitrary cost parameters (from untrusted sources)
+	// since large cost parameters can exhaust server resources.
+	const (
+		MaxTimeCost   = 10         // At most 10 passes over the memory
+		MaxMemoryCost = 256 * 1024 // At most 256 MiB memory
+		MaxThreads    = 16         // At most 16 threads
+	)
+
+	if params.Algorithm != "Argon2id" {
+		return nil, errors.New("config: invalid key derivation algorithm")
+	}
+	if params.TimeCost < 1 || params.TimeCost > MaxTimeCost {
+		return nil, errors.New("config: invalid key derivation time cost")
+	}
+	if params.MemoryCost > MaxMemoryCost {
+		return nil, errors.New("config: invalid key derivation memory cost")
+	}
+	if params.Threads < 1 || params.Threads > MaxThreads {
+		return nil, errors.New("config: invalid key derivation parallelism")
+	}
+	return MasterKey(argon2.IDKey(
+		[]byte(password),
+		params.Salt,
+		uint32(params.TimeCost),
+		uint32(params.MemoryCost),
+		uint8(params.Threads),
+		32,
+	)), nil
+}
+
+// EncryptBytes encrypts and authenticates the given data
+// and returns the encrypted ciphertext.
+func (key MasterKey) EncryptBytes(data []byte) ([]byte, error) {
+	ciphertext := bytes.NewBuffer(make([]byte, 0, len(data)))
+
+	w, err := key.Encrypt(ciphertext)
+	if err != nil {
+		return nil, err
+	}
+	if _, err = io.Copy(w, bytes.NewReader(data)); err != nil {
+		return nil, err
+	}
+	if err = w.Close(); err != nil {
+		return nil, err
+	}
+	return ciphertext.Bytes(), nil
+}
+
+// Encrypt returns a new io.WriterCloser that wraps w and
+// encrypts and authenticates everything written to it.
+//
+// The returned io.WriteCloser has to be closed successfully to
+// complete the encryption process.
+func (key MasterKey) Encrypt(w io.Writer) (io.WriteCloser, error) {
+	iv, err := sioutil.Random(16)
+	if err != nil {
+		return nil, err
+	}
+
+	prf := hmac.New(sha256.New, key)
+	prf.Write(iv)
+	fileKey := prf.Sum(nil)
+
+	nonce, err := sioutil.Random(8)
+	if err != nil {
+		return nil, err
+	}
+
+	const (
+		AES256GCM        = 0
+		ChaCha20Poly1305 = 1
+	)
+	var (
+		algorithm   sio.Algorithm
+		algorithmID [4]byte
+	)
+	if sioutil.NativeAES() {
+		algorithm = sio.AES_256_GCM
+		binary.LittleEndian.PutUint32(algorithmID[:], AES256GCM)
```

Comment:
My efficiency bones are tingling. Glad you want to support 2^32 algorithms :D

Could we make the top byte a version for the entire package?

`const version = 1<<24`
```suggestion
		binary.LittleEndian.PutUint32(algorithmID[:], AES256GCM | version)
```

You can also slap it in front as a single byte and *consider* just making this a single byte :)

The decoder should just check that it is 1 and reject it if not, so downgrades will not have side effects.

Reply:
> My efficiency bones are tingling. Glad you want to support 2^32 algorithms :D

Yeah, `2³²` algorithms is way to much. W.r.t. the `1` byte suggestion - I did that before, see: https://github.com/minio/minio/blob/master/pkg/madmin/encrypt.go#L93

My reasoning here was: Align the prefix: `4 + 16 + 8` bytes. Also the `4` bytes could be split in the future - like you're suggesting with the version number.

> Could we make the top byte a version for the entire package?

We can - most likely an update will not be required since the scheme is not "specific" for the config file context and the encryption scheme has a security proof, but it is a cleaner and more structured representation.

Reply:
Updated the scheme to use a 16 bit header split into `<version> | <algorithm-id>`

Reply:
Alignment is not really that much a concern in serialized data since you are typically memcopying anyway, and futhermore unaligned accesses are not slower on CPUs released in the last 5-10 years.

The slicing, bounds check and everything needed for `binary.LittleEndian.Uint32` is much bigger than just copying a byte.

> most likely an update will not be required since the scheme is now "specific"

I've heard that many times before ;)